File: scripts/day10-part2-simple.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <regex>
#include <chrono>
#include <map>
#include <bitset>

using namespace std;

struct Machine {
    string lights;
    vector<vector<int>> buttons;
    vector<int> joltages;
};

Machine parseMachine(const string& line) {
    Machine m;
    regex pattern(R"(\[(.*?)\](.*)\{(.*)\})");
    smatch match;
    regex_search(line, match, pattern);
    
    m.lights = match[1];
    string buttonsStr = match[2];
    string joltagesStr = match[3];
    
    size_t pos = 0;
    while ((pos = joltagesStr.find(',')) != string::npos) {
        m.joltages.push_back(stoi(joltagesStr.substr(0, pos)));
        joltagesStr.erase(0, pos + 1);
    }
    m.joltages.push_back(stoi(joltagesStr));
    
    regex buttonPattern(R"(\(([0-9,]+)\))");
    sregex_iterator iter(buttonsStr.begin(), buttonsStr.end(), buttonPattern);
    sregex_iterator end;
    
    while (iter != end) {
        vector<int> button;
        string btnStr = (*iter)[1];
        size_t p = 0;
        while ((p = btnStr.find(',')) != string::npos) {
            button.push_back(stoi(btnStr.substr(0, p)));
            btnStr.erase(0, p + 1);
        }
        button.push_back(stoi(btnStr));
        m.buttons.push_back(button);
        ++iter;
    }
    
    return m;
}

map<vector<int>, map<vector<int>, int>> getAllParityMaps(int nCounters, const vector<vector<int>>& switches) {
    map<vector<int>, map<vector<int>, int>> parityMaps;
    
    for (int i = 0; i < (1 << switches.size()); i++) {
        bitset<16> bits(i);
        vector<int> result(nCounters, 0);
        
        for (int j = 0; j < switches.size(); j++) {
            if (bits[j]) {
                for (int ele : switches[j]) {
                    result[ele]++;
                }
            }
        }
        
        vector<int> parityKey;
        for (int ele : result) {
            parityKey.push_back(ele % 2);
        }
        
        int nFlips = bits.count();
        
        if (parityMaps[parityKey].find(result) == parityMaps[parityKey].end() ||
            parityMaps[parityKey][result] > nFlips) {
            parityMaps[parityKey][result] = nFlips;
        }
    }
    
    return parityMaps;
}

pair<bool, int> findMinSwitchFlips(
    const vector<int>& current,
    const vector<vector<int>>& switches,
    const map<vector<int>, map<vector<int>, int>>& parityMaps,
    map<vector<int>, pair<bool, int>>& cache
) {
    if (cache.find(current) != cache.end()) {
        return cache[current];
    }
    
    bool allZero = true;
    for (int x : current) {
        if (x != 0) {
            allZero = false;
            break;
        }
    }
    if (allZero) {
        cache[current] = {true, 0};
        return {true, 0};
    }
    
    for (int x : current) {
        if (x < 0) {
            cache[current] = {false, 0};
            return {false, 0};
        }
    }
    
    vector<int> currentParity;
    for (int c : current) {
        currentParity.push_back(c % 2);
    }
    
    if (parityMaps.find(currentParity) == parityMaps.end()) {
        cache[current] = {false, 0};
        return {false, 0};
    }
    
    int minFlips = 999999;
    
    for (const auto& [pattern, nFlips] : parityMaps.at(currentParity)) {
        bool valid = true;
        for (int i = 0; i < pattern.size(); i++) {
            if (pattern[i] > current[i]) {
                valid = false;
                break;
            }
        }
        
        if (!valid) continue;
        
        vector<int> next;
        for (int i = 0; i < current.size(); i++) {
            next.push_back((current[i] - pattern[i]) / 2);
        }
        
        auto [found, count] = findMinSwitchFlips(next, switches, parityMaps, cache);
        
        if (found) {
            minFlips = min(minFlips, nFlips + 2 * count);
        }
    }
    
    if (minFlips == 999999) {
        cache[current] = {false, 0};
        return {false, 0};
    }
    
    cache[current] = {true, minFlips};
    return {true, minFlips};
}

int solvePart2(const vector<Machine>& machines) {
    int total = 0;
    
    for (const auto& machine : machines) {
        auto parityMaps = getAllParityMaps(machine.joltages.size(), machine.buttons);
        map<vector<int>, pair<bool, int>> cache;
        auto [found, count] = findMinSwitchFlips(machine.joltages, machine.buttons, parityMaps, cache);
        total += count;
    }
    
    return total;
}

int main() {
    ifstream file("../inputs/day-10-input.txt");
    vector<Machine> machines;
    string line;
    
    while (getline(file, line)) {
        machines.push_back(parseMachine(line));
    }
    file.close();
    
    const int iterations = 1;
    auto start = chrono::high_resolution_clock::now();
    
    int result;
    for (int i = 0; i < iterations; i++) {
        result = solvePart2(machines);
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::nanoseconds>(end - start);
    double avgMicros = duration.count() / (double)iterations / 1000.0;
    
    cout << "C++ (Day 10 Part 2 - Simple)" << endl;
    cout << "Answer: " << result << endl;
    cout << "Average time: " << avgMicros << " microseconds (" << iterations << " iterations)" << endl;
    cout << "Total time: " << (duration.count() / 1000000.0) << " ms" << endl;
    
    return 0;
}

File: scripts/day10-part2-simple.py
import time
import re

def parse_machine(line):
    match = re.match(r'\[(.*?)\](.*)\{(.*)\}', line)
    lights = match.group(1)
    buttons_str = match.group(2).strip()
    joltages = list(map(int, match.group(3).split(',')))
    
    buttons = []
    for button_match in re.finditer(r'\(([0-9,]+)\)', buttons_str):
        buttons.append(list(map(int, button_match.group(1).split(','))))
    
    return {'lights': lights, 'buttons': buttons, 'joltages': joltages}

def get_all_parity_maps(n_counters, switches):
    parity_maps = {}
    
    for i in range(1 << len(switches)):
        result = [0] * n_counters
        n_flips = 0
        
        for j in range(len(switches)):
            if i & (1 << j):
                n_flips += 1
                for ele in switches[j]:
                    result[ele] += 1
        
        parity_key = tuple(x % 2 for x in result)
        result_key = tuple(result)
        
        if parity_key not in parity_maps:
            parity_maps[parity_key] = {}
        
        if result_key not in parity_maps[parity_key] or \
           parity_maps[parity_key][result_key] > n_flips:
            parity_maps[parity_key][result_key] = n_flips
    
    return parity_maps

def find_min_switch_flips(current, switches, parity_maps, cache):
    key = tuple(current)
    
    if key in cache:
        return cache[key]
    
    if all(x == 0 for x in current):
        cache[key] = (True, 0)
        return (True, 0)
    
    if any(x < 0 for x in current):
        cache[key] = (False, 0)
        return (False, 0)
    
    current_parity = tuple(x % 2 for x in current)
    
    if current_parity not in parity_maps:
        cache[key] = (False, 0)
        return (False, 0)
    
    min_flips = float('inf')
    
    for pattern, n_flips in parity_maps[current_parity].items():
        # Check validity
        if not all(p <= c for p, c in zip(pattern, current)):
            continue
        
        next_state = tuple((c - p) // 2 for c, p in zip(current, pattern))
        found, count = find_min_switch_flips(list(next_state), switches, parity_maps, cache)
        
        if found:
            min_flips = min(min_flips, n_flips + 2 * count)
    
    if min_flips == float('inf'):
        cache[key] = (False, 0)
        return (False, 0)
    
    cache[key] = (True, min_flips)
    return (True, min_flips)

def solve_part2(machines):
    total = 0
    
    for machine in machines:
        parity_maps = get_all_parity_maps(len(machine['joltages']), machine['buttons'])
        cache = {}
        found, count = find_min_switch_flips(machine['joltages'], machine['buttons'], parity_maps, cache)
        total += count
    
    return total

# Read input
with open("../inputs/day-10-input.txt", "r") as f:
    input_data = f.read().strip().split('\n')

machines = [parse_machine(line) for line in input_data]

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part2(machines)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 10 Part 2 - Simple)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")
